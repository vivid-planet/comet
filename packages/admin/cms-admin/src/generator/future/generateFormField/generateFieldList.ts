import { IntrospectionField, IntrospectionQuery, IntrospectionType } from "graphql";

export function generateFieldListFromIntrospection(
    gqlIntrospection: IntrospectionQuery,
    type: string,
): { path: string; field: IntrospectionField }[] {
    return fieldListFromIntrospectionTypeRecursive(gqlIntrospection.__schema.types, type);
}

function fieldListFromIntrospectionTypeRecursive(
    types: readonly IntrospectionType[],
    type: string,
    parentPath?: string,
): { path: string; field: IntrospectionField }[] {
    const typeDef = types.find((typeDef) => typeDef.name === type);
    if (!typeDef || typeDef.kind !== "OBJECT") return [];

    return typeDef.fields.reduce<{ path: string; field: IntrospectionField }[]>((acc, field) => {
        const path = `${parentPath ? `${parentPath}.` : ""}${field.name}`;
        let type = field.type;
        if (type.kind === "NON_NULL") type = type.ofType;

        if (type.kind === "OBJECT") {
            acc.push({ path: path, field: field });
            acc.push(...fieldListFromIntrospectionTypeRecursive(types, type.name, path));
        } else {
            acc.push({ path: path, field: field });
        }
        return acc;
    }, []);
}
