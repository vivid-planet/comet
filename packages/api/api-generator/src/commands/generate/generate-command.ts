import { type CrudGeneratorOptions, type CrudSingleGeneratorOptions } from "@comet/cms-api";
import { CLIHelper } from "@mikro-orm/cli";
import { LazyMetadataStorage } from "@nestjs/graphql/dist/schema-builder/storages/lazy-metadata.storage";
import { Command } from "commander";

import { generateCrud } from "./generateCrud/generate-crud";
import { generateCrudSingle } from "./generateCrudSingle/generate-crud-single";
import { writeGeneratedFiles } from "./utils/write-generated-files";

export const generateCommand = new Command("generate").action(async (options) => {
    try {
        const orm = await CLIHelper.getORM(undefined, undefined, { dbName: "generator" });

        const entities = orm.em.getMetadata().getAll();
        LazyMetadataStorage.load();

        for (const name in entities) {
            const entity = entities[name];
            if (!entity.class) {
                // Ignore e.g. relation entities that don't have a class
                continue;
            }
            {
                const generatorOptions = Reflect.getMetadata(`data:crudGeneratorOptions`, entity.class) as CrudGeneratorOptions | undefined;
                if (generatorOptions) {
                    const files = await generateCrud(generatorOptions, entity);
                    await writeGeneratedFiles(files, { targetDirectory: generatorOptions.targetDirectory });
                }
            }
            {
                const generatorOptions = Reflect.getMetadata(`data:crudSingleGeneratorOptions`, entity.class) as
                    | CrudSingleGeneratorOptions
                    | undefined;
                if (generatorOptions) {
                    const files = await generateCrudSingle(generatorOptions, entity);
                    await writeGeneratedFiles(files, { targetDirectory: generatorOptions.targetDirectory });
                }
            }
        }
        await orm.close(true);
    } catch (e) {
        console.warn(e);
    }
});
