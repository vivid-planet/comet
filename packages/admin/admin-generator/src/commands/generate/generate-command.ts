/* eslint-disable @typescript-eslint/no-explicit-any */
import { type GridColDef } from "@comet/admin";
import { type IconName } from "@comet/admin-icons";
import { type BlockInterface, type FinalFormFileUploadProps } from "@comet/cms-admin";
import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import { loadSchema } from "@graphql-tools/load";
import { type IconProps } from "@mui/material";
import {
    type GridCellParams,
    type GridFilterItem,
    type GridFilterOperator,
    type GridRenderCellParams,
    type GridSortDirection,
    type GridValidRowModel,
} from "@mui/x-data-grid";
import { Command } from "commander";
import { type FieldValidator } from "final-form";
import { promises as fs } from "fs";
import { glob } from "glob";
import { introspectionFromSchema } from "graphql";
import { basename, dirname } from "path";
import { type ComponentType } from "react";

import { parseConfig } from "./config/parseConfig";
import { generateForm } from "./generateForm/generateForm";
import { generateGrid } from "./generateGrid/generateGrid";
import { type UsableFields } from "./generateGrid/usableFields";
import { type ColumnVisibleOption } from "./utils/columnVisibility";
import { writeGenerated } from "./utils/writeGenerated";

type IconObject = Pick<IconProps, "color" | "fontSize"> & {
    name: IconName;
};

type Icon = IconName | IconObject | ComponentType;
export type Adornment = string | { icon: Icon };

type InputBaseFieldConfig = {
    startAdornment?: Adornment;
    endAdornment?: Adornment;
};

export type ComponentFormFieldConfig = { type: "component"; component: ComponentType };
function isComponentFormFieldConfig(arg: any): arg is ComponentFormFieldConfig {
    return arg && arg.type === "component";
}

export type StaticSelectValue = { value: string; label: string } | string;

export type FormFieldConfig<T> = (
    | ({ type: "text"; name: keyof T; multiline?: boolean } & InputBaseFieldConfig)
    | ({ type: "number"; name: keyof T; decimals?: number } & InputBaseFieldConfig)
    | ({
          type: "numberRange";
          name: keyof T;
          minValue: number;
          maxValue: number;
          disableSlider?: boolean;
      } & InputBaseFieldConfig)
    | { type: "boolean"; name: keyof T }
    | ({ type: "date"; name: keyof T } & InputBaseFieldConfig)
    | ({ type: "dateTime"; name: keyof T } & InputBaseFieldConfig)
    | ({
          type: "staticSelect";
          name: keyof T;
          values?: StaticSelectValue[];
          inputType?: "select" | "radio";
      } & Omit<InputBaseFieldConfig, "endAdornment">)
    | ({
          type: "asyncSelect";
          name: keyof T;
          rootQuery: string;
          labelField?: string;
          /**
           * filter for query, passed as variable to graphql query
           */
          filter?:
              | {
                    /**
                     * Filter by value of field in current form
                     */
                    type: "field";
                    /**
                     * Name of the field in current form, that will be used to filter the query
                     */
                    formFieldName: string;
                    /**
                     * Name of the graphql argument the prop will be applied to. Defaults to propdName.
                     *
                     * Root Argument or filter argument are supported.
                     */
                    rootQueryArg?: string;
                }
              | {
                    /**
                     * Filter by a prop passed into the form, this prop will be generated
                     */
                    type: "formProp";
                    /**
                     * Name of the prop generated for this form
                     */
                    propName: string;
                    /**
                     * Name of the graphql argument the prop will be applied to. Defaults to propdName.
                     *
                     * Root Argument or filter argument are supported.
                     */
                    rootQueryArg?: string;
                };
      } & Omit<InputBaseFieldConfig, "endAdornment">)
    | ({
          type: "asyncSelectFilter";
          name: string;
          loadValueQueryField: string; //TODO improve typing, use something similar to UsableFields<T>;
          rootQuery: string;
          labelField?: string;
      } & Omit<InputBaseFieldConfig, "endAdornment">)
    | { type: "block"; name: keyof T; block: BlockInterface }
    | ({ type: "fileUpload"; multiple?: false; name: keyof T; maxFiles?: 1; download?: boolean } & Pick<
          Partial<FinalFormFileUploadProps<false>>,
          "maxFileSize" | "readOnly" | "layout" | "accept"
      >)
    | ({ type: "fileUpload"; multiple: true; name: keyof T; maxFiles?: number; download?: boolean } & Pick<
          Partial<FinalFormFileUploadProps<true>>,
          "maxFileSize" | "readOnly" | "layout" | "accept"
      >)
) & {
    label?: string;
    required?: boolean;
    virtual?: boolean;
    validate?: FieldValidator<unknown>;
    helperText?: string;
    readOnly?: boolean;
};

export function isFormFieldConfig<T>(arg: any): arg is FormFieldConfig<T> {
    return !isFormLayoutConfig(arg) && !isComponentFormFieldConfig(arg);
}

type OptionalNestedFieldsConfig<T> = {
    type: "optionalNestedFields";
    name: keyof T; // object name containing fields
    checkboxLabel?: string;
    fields: FormFieldConfig<any>[];
};
export type FormLayoutConfig<T> =
    | {
          type: "fieldSet";
          name: string;
          title?: string;
          supportText?: string; // can contain field-placeholder
          collapsible?: boolean; // default true
          initiallyExpanded?: boolean; // default false
          fields: (FormFieldConfig<T> | OptionalNestedFieldsConfig<T> | ComponentFormFieldConfig)[];
      }
    | OptionalNestedFieldsConfig<T>;

export function isFormLayoutConfig<T>(arg: any): arg is FormLayoutConfig<T> {
    return arg.type !== undefined && ["fieldSet", "optionalNestedFields"].includes(arg.type);
}

export type FormConfig<T extends { __typename?: string }> = {
    type: "form";
    gqlType: T["__typename"];
    mode?: "edit" | "add" | "all";
    fragmentName?: string;
    createMutation?: string;
    fields: (FormFieldConfig<T> | FormLayoutConfig<T> | ComponentFormFieldConfig)[];
};

type BaseColumnConfig = Pick<GridColDef, "headerName" | "width" | "minWidth" | "maxWidth" | "flex" | "pinned" | "disableExport"> & {
    headerInfoTooltip?: string;
    visible?: ColumnVisibleOption;
    fieldName?: string; // this can be used to overwrite field-prop of column-config
};

export type GridColumnStaticSelectLabelCellContent = {
    primaryText?: string;
    secondaryText?: string;
    icon?: Icon;
};

export type GridColumnStaticSelectValue =
    | StaticSelectValue
    | {
          value: string | number | boolean;
          label: string | GridColumnStaticSelectLabelCellContent;
      }
    | number
    | boolean;

export type GridColumnConfig<T extends GridValidRowModel> = (
    | { type: "text"; renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element }
    | { type: "number"; currency?: string; decimals?: number; renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element }
    | { type: "boolean"; renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element }
    | { type: "date"; renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element }
    | { type: "dateTime"; renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element }
    | { type: "staticSelect"; values?: GridColumnStaticSelectValue[] }
    | { type: "block"; block: BlockInterface }
    | { type: "id"; renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element }
    | {
          type: "manyToMany";
          renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element;
          queryFields?: UsableFields<T, true>[];
          /**
           * The field to use as label for the default renderCell implementation.
           */
          labelField?: string;
      }
    | {
          type: "oneToMany";
          renderCell?: (params: GridRenderCellParams<T, any, any>) => JSX.Element;
          queryFields?: UsableFields<T, true>[];
          /**
           * The field to use as label for the default renderCell implementation.
           */
          labelField?: string;
      }
) & { name: UsableFields<T>; filterOperators?: GridFilterOperator[] } & BaseColumnConfig;

export type ActionsGridColumnConfig<T> = {
    type: "actions";
    queryFields?: UsableFields<T, true>[];
    component?: ComponentType<GridCellParams>;
} & BaseColumnConfig;
export type VirtualGridColumnConfig<T extends GridValidRowModel> = {
    type: "virtual";
    name: string;
    queryFields?: UsableFields<T, true>[];
    renderCell: (params: GridRenderCellParams<T, any, any>) => JSX.Element;
} & Pick<GridColDef, "sortBy"> &
    BaseColumnConfig;

type InitialFilterConfig = {
    items: GridFilterItem[];
    linkOperator?: "and" | "or";
};

// Additional type is necessary to avoid "TS2589: Type instantiation is excessively deep and possibly infinite."
type GridConfigGridColumnDef<T extends { __typename?: string }> = GridColumnConfig<T> | ActionsGridColumnConfig<T> | VirtualGridColumnConfig<T>;

export type GridConfig<T extends { __typename?: string }> = {
    type: "grid";
    gqlType: T["__typename"];
    fragmentName?: string;
    query?: string;
    queryParamsPrefix?: string;
    columns: Array<GridConfigGridColumnDef<T>>;
    excelExport?: boolean;
    add?: boolean;
    edit?: boolean;
    delete?: boolean;
    readOnly?: boolean;
    initialSort?: Array<{ field: string; sort: GridSortDirection }>;
    initialFilter?: InitialFilterConfig;
    filterProp?: boolean;
    toolbar?: boolean;
    toolbarActionProp?: boolean;
    newEntryText?: string;
    rowActionProp?: boolean;
    selectionProps?: "multiSelect" | "singleSelect";
    rowReordering?: {
        enabled: boolean;
        dragPreviewField?: UsableFields<T>;
    };
};

export type GeneratorConfig<T extends { __typename?: string }> = FormConfig<T> | GridConfig<T>;

export function defineConfig<T extends { __typename?: string }>(config: GeneratorConfig<T>) {
    return config;
}

type GQLDocumentConfig = { document: string; export: boolean };
export type GQLDocumentConfigMap = Record<string, GQLDocumentConfig>;
export type GeneratorReturn = { code: string; gqlDocuments: GQLDocumentConfigMap };

/**
 * @experimental
 */
async function runGenerate(filePattern = "src/**/*.cometGen.{ts,tsx}") {
    const schema = await loadSchema("./schema.gql", {
        loaders: [new GraphQLFileLoader()],
    });
    const gqlIntrospection = introspectionFromSchema(schema);

    const files: string[] = await glob(filePattern);
    for (const file of files) {
        let outputCode = "";
        let gqlDocumentsOutputCode = "";
        const targetDirectory = `${dirname(file)}/generated`;
        const baseOutputFilename = basename(file).replace(/\.cometGen\.tsx?$/, "");

        console.log(`generating ${file}`);

        const config = await parseConfig(file);

        const codeOuputFilename = `${targetDirectory}/${basename(file.replace(/\.cometGen\.tsx?$/, ""))}.tsx`;
        await fs.rm(codeOuputFilename, { force: true });

        const exportName = file.match(/([^/]+)\.cometGen\.tsx?$/)?.[1];
        if (!exportName) throw new Error("Can not determine exportName");

        let generated: GeneratorReturn;
        if (config.type == "form") {
            generated = generateForm({ exportName, gqlIntrospection, baseOutputFilename, targetDirectory }, config);
        } else if (config.type == "grid") {
            generated = generateGrid({ exportName, gqlIntrospection, baseOutputFilename, targetDirectory }, config);
        } else {
            throw new Error(`Unknown config type`);
        }
        outputCode += generated.code;
        for (const queryName in generated.gqlDocuments) {
            const exportStatement = generated.gqlDocuments[queryName].export ? "export " : "";
            gqlDocumentsOutputCode += `${exportStatement} const ${queryName} = gql\`${generated.gqlDocuments[queryName].document}\`\n`;
        }

        await writeGenerated(codeOuputFilename, outputCode);

        if (gqlDocumentsOutputCode != "") {
            const gqlDocumentsOuputFilename = `${targetDirectory}/${basename(file.replace(/\.cometGen\.tsx?$/, ""))}.gql.tsx`;
            await fs.rm(gqlDocumentsOuputFilename, { force: true });
            gqlDocumentsOutputCode = `import { gql } from "@apollo/client";
                import { finalFormFileUploadFragment, finalFormFileUploadDownloadableFragment } from "@comet/cms-admin";

                ${gqlDocumentsOutputCode}
            `;
            await writeGenerated(gqlDocumentsOuputFilename, gqlDocumentsOutputCode);
        }
        console.log("");
    }
}

export const generateCommand = new Command("generate")
    .option("-f, --file <file>", "path to config file or glob pattern to generate specific files")
    .action(async ({ file: filePattern }: { file?: string }) => {
        console.log("️️️⚠️️️⚠️️️⚠️️️  Admin Generator is still experimental and in beta phase.  ⚠️️️⚠️️️⚠️️️\n\n");
        await runGenerate(filePattern);
    });
