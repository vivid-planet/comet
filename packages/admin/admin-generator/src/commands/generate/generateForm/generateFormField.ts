import { type IntrospectionEnumType, type IntrospectionNamedTypeRef, type IntrospectionQuery } from "graphql";

import { type FormConfig, type FormFieldConfig, type GQLDocumentConfigMap } from "../generate-command";
import { camelCaseToHumanReadable } from "../utils/camelCaseToHumanReadable";
import { convertConfigImport } from "../utils/convertConfigImport";
import { type Imports } from "../utils/generateImportsCode";
import { isFieldOptional } from "../utils/isFieldOptional";
import { isGeneratorConfigCode, isGeneratorConfigImport } from "../utils/runtimeTypeGuards";
import { generateAsyncSelect } from "./asyncSelect/generateAsyncSelect";
import { buildFormFieldOptions } from "./formField/options";
import { type GenerateFieldsReturn } from "./generateFields";
import { type Prop } from "./generateForm";

export function generateFormField({
    gqlIntrospection,
    baseOutputFilename,
    config,
    formConfig,
    gqlType,
    namePrefix,
}: {
    gqlIntrospection: IntrospectionQuery;
    baseOutputFilename: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: FormFieldConfig<any>;
    formFragmentName: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    formConfig: FormConfig<any>;
    gqlType: string;
    namePrefix?: string;
}): GenerateFieldsReturn {
    if (config.type == "asyncSelect" || config.type == "asyncSelectFilter") {
        return generateAsyncSelect({ gqlIntrospection, baseOutputFilename, config, formConfig, gqlType, namePrefix });
    }
    const imports: Imports = [];
    const formProps: Prop[] = [];

    const {
        name,
        formattedMessageRootId,
        fieldLabel,
        startAdornment,
        endAdornment,
        imports: optionsImports,
    } = buildFormFieldOptions({ config, formConfig, gqlType, gqlIntrospection });
    imports.push(...optionsImports);

    let { introspectionFieldType } = buildFormFieldOptions({ config, formConfig, gqlType, gqlIntrospection });

    const nameWithPrefix = `${namePrefix ? `${namePrefix}.` : ``}${name}`;

    const rootGqlType = formConfig.gqlType;

    const dataRootName = rootGqlType[0].toLowerCase() + rootGqlType.substring(1); // TODO should probably be deteced via query

    const required = !isFieldOptional({ config, gqlIntrospection, gqlType });

    //TODO verify introspectionField.type is compatbile with config.type

    const endAdornmentWithLockIconProp = `endAdornment={<InputAdornment position="end"><Lock /></InputAdornment>}`;
    const readOnlyProps = `readOnly disabled`;
    const readOnlyPropsWithLock = `${readOnlyProps} ${endAdornmentWithLockIconProp}`;

    let formValuesConfig: GenerateFieldsReturn["formValuesConfig"] = [{}]; // FormFields should only contain one entry

    const gqlDocuments: GQLDocumentConfigMap = {};
    const hooksCode = "";
    let finalFormConfig: GenerateFieldsReturn["finalFormConfig"];

    let validateCode = "";
    if (config.validate) {
        if (isGeneratorConfigImport(config.validate)) {
            imports.push(convertConfigImport(config.validate));
            validateCode = `validate={${config.validate.name}}`;
        } else if (isGeneratorConfigCode(config.validate)) {
            validateCode = `validate={${config.validate.code}}`;
            imports.push(...config.validate.imports.map((imprt) => convertConfigImport(imprt)));
        }
    }

    let code = "";
    let formValueToGqlInputCode = "";
    let formFragmentFields = [name];
    if (config.type == "text") {
        const TextInputComponent = config.multiline ? "TextAreaField" : "TextField";
        code = `
        <${TextInputComponent}
            ${required ? "required" : ""}
            ${config.readOnly ? readOnlyPropsWithLock : ""}
            variant="horizontal"
            fullWidth
            name="${nameWithPrefix}"
            label={${fieldLabel}}
            ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
            ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` +
                      `"${formattedMessageRootId}.${name}.helperText" ` +
                      `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
        />`;
        if (!required && !config.readOnly) {
            formValueToGqlInputCode = `${name}: formValues.${name} ?? null,`;
        }
    } else if (config.type == "number") {
        code = `
            <NumberField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.decimals ? `decimals={${config.decimals}}` : ""}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        //TODO MUI suggest not using type=number https://mui.com/material-ui/react-text-field/#type-quot-number-quot
        let assignment = `parseFloat(formValues.${nameWithPrefix})`;
        if (isFieldOptional({ config, gqlIntrospection: gqlIntrospection, gqlType: gqlType })) {
            assignment = `formValues.${nameWithPrefix} ? ${assignment} : null`;
        }
        formValueToGqlInputCode = `${name}: ${assignment},`;

        let initializationAssignment = `String(data.${dataRootName}.${nameWithPrefix})`;
        if (!required) {
            initializationAssignment = `data.${dataRootName}.${nameWithPrefix} ? ${initializationAssignment} : undefined`;
        }
        formValuesConfig = [
            {
                omitFromFragmentType: name,
                typeCode: `${name}${!required ? `?` : ``}: string;`,
                initializationCode: `${name}: ${initializationAssignment}`,
            },
        ];
    } else if (config.type === "numberRange") {
        code = `
            <Field
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                component={FinalFormRangeInput}
                label={${fieldLabel}}
                min={${config.minValue}}
                max={${config.maxValue}}
                ${config.disableSlider ? "disableSlider" : ""}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;

        formFragmentFields = [`${name}.min`, `${name}.max`];
    } else if (config.type == "boolean") {
        code = `<CheckboxField
                        label={${fieldLabel}}
                        name="${nameWithPrefix}"
                        fullWidth
                        variant="horizontal"
                        ${config.readOnly ? readOnlyProps : ""}
                        ${
                            config.helperText
                                ? `helperText={<FormattedMessage id=` +
                                  `"${formattedMessageRootId}.${name}.helperText" ` +
                                  `defaultMessage="${config.helperText}" />}`
                                : ""
                        }
                        ${validateCode}
                    />`;
        formValuesConfig = [
            {
                defaultInitializationCode: `${name}: false`,
            },
        ];
    } else if (config.type == "date") {
        imports.push({
            name: "Future_DatePickerField",
            importPath: "@comet/admin",
        });
        code = `
            <Future_DatePickerField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        if (!required && !config.readOnly) {
            formValueToGqlInputCode = `${name}: formValues.${name} ?? null,`;
        }
    } else if (config.type == "dateTime") {
        code = `<DateTimeField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        formValuesConfig = [
            {
                initializationCode: `${name}: data.${dataRootName}.${nameWithPrefix} ? new Date(data.${dataRootName}.${nameWithPrefix}) : undefined`,
                omitFromFragmentType: name,
                typeCode: `${name}${!required ? "?" : ""}: Date${!required ? " | null" : ""};`,
            },
        ];
        if (!config.readOnly) {
            formValueToGqlInputCode = required
                ? `${name}: formValues.${name}.toISOString(),`
                : `${name}: formValues.${name} ? formValues.${name}.toISOString() : null,`;
        }
    } else if (config.type == "block") {
        code = `<Field name="${nameWithPrefix}" isEqual={isEqual} label={${fieldLabel}} variant="horizontal" fullWidth>
            {createFinalFormBlock(rootBlocks.${String(config.name)})}
        </Field>`;
        formValueToGqlInputCode = `${name}: rootBlocks.${name}.state2Output(formValues.${nameWithPrefix}),`;
        formValuesConfig = [
            {
                typeCode: `${name}: BlockState<typeof rootBlocks.${name}>;`,
                initializationCode: `${name}: rootBlocks.${name}.input2State(data.${dataRootName}.${nameWithPrefix})`,
                defaultInitializationCode: `${name}: rootBlocks.${name}.defaultValues()`,
            },
        ];
    } else if (config.type === "fileUpload") {
        const multiple = config.multiple || (typeof config.maxFiles === "number" && config.maxFiles > 1);
        code = `<FileUploadField name="${name}" label={${fieldLabel}}
            variant="horizontal"
            ${config.multiple ? "multiple" : ""}
            ${config.maxFiles ? `maxFiles={${config.maxFiles}}` : ""}
            ${config.maxFileSize ? `maxFileSize={${config.maxFileSize}}` : ""}
            ${config.readOnly ? `readOnly` : ""}
            ${config.layout ? `layout="${config.layout}"` : ""}
            ${config.accept ? `accept="${config.accept}"` : ""}
        />`;
        if (multiple) {
            formValueToGqlInputCode = `${name}: formValues.${name}?.map(({ id }) => id),`;
        } else {
            formValueToGqlInputCode = `${name}: formValues.${name} ? formValues.${name}.id : null,`;
        }
        formFragmentFields = [`${name} { ...${config.download ? "FinalFormFileUploadDownloadable" : "FinalFormFileUpload"} }`];
    } else if (config.type == "staticSelect") {
        const multiple = introspectionFieldType?.kind === "LIST";
        if (introspectionFieldType?.kind === "LIST") {
            introspectionFieldType =
                introspectionFieldType.ofType.kind === "NON_NULL" ? introspectionFieldType.ofType.ofType : introspectionFieldType.ofType;
        }

        const enumType = gqlIntrospection.__schema.types.find(
            (t) => t.kind === "ENUM" && t.name === (introspectionFieldType as IntrospectionNamedTypeRef).name,
        ) as IntrospectionEnumType | undefined;
        if (!enumType) throw new Error(`Enum type ${(introspectionFieldType as IntrospectionNamedTypeRef).name} not found for field ${name}`);

        const values = (config.values ? config.values : enumType.enumValues.map((i) => i.name)).map((value) => {
            if (typeof value === "string") {
                return {
                    value,
                    label: camelCaseToHumanReadable(value),
                };
            } else {
                return value;
            }
        });

        let inputType = config.inputType;
        if (!inputType) {
            if (!required || multiple) {
                // radio is not clearable, render always as select if not required
                // radio doesn't support multiple
                inputType = "select";
            } else {
                // auto select/radio based on number of values
                if (values.length <= 5) {
                    inputType = "radio";
                } else {
                    inputType = "select";
                }
            }
        }
        if (inputType === "radio" && multiple) {
            throw new Error(`${name}: inputType=radio doesn't support multiple`);
        }
        if (inputType === "radio" && !required) {
            throw new Error(`${name}: inputType=radio must be required as it doesn't support clearable`);
        }

        const renderAsRadio = inputType === "radio";
        if (renderAsRadio) {
            code = `<RadioGroupField
             ${required ? "required" : ""}
              variant="horizontal"
             fullWidth
             name="${nameWithPrefix}"
             label={${fieldLabel}}
             options={[
                  ${values
                      .map((value) => {
                          return `{
                                label: <FormattedMessage id="${formattedMessageRootId}.${name}.${
                                    value.value.charAt(0).toLowerCase() + value.value.slice(1)
                                }" defaultMessage="${value.label}" />,
                                value: "${value.value}",
                            }`;
                      })
                      .join(",")}
            ]}/>`;
        } else {
            imports.push({
                name: "SelectField",
                importPath: "@comet/admin",
            });
            code = `<SelectField
                            ${required ? "required" : ""}
                             fullWidth
                                variant={"horizontal"}
                                name="${nameWithPrefix}"
                                label={${fieldLabel}}
                                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                                ${
                                    config.helperText
                                        ? `helperText={<FormattedMessage id=` +
                                          `"${formattedMessageRootId}.${name}.helperText" ` +
                                          `defaultMessage="${config.helperText}" />}`
                                        : ""
                                }
                                ${validateCode}
                                ${config.readOnly ? readOnlyPropsWithLock : ""}
                                ${multiple ? "multiple" : ""}
                                options={[${values.map((value) => {
                                    const id = `${formattedMessageRootId}.${name}.${value.value.charAt(0).toLowerCase() + value.value.slice(1)}`;

                                    return `{
                                        value: "${value.value}",
                                        label: <FormattedMessage id="${id}" defaultMessage="${value.label}" />
                                    }`;
                                })}]}
                            />`;
        }
    } else {
        throw new Error(`Unsupported type`);
    }
    return {
        code,
        hooksCode,
        formValueToGqlInputCode,
        formFragmentFields,
        gqlDocuments,
        imports,
        formProps,
        formValuesConfig,
        finalFormConfig,
    };
}
