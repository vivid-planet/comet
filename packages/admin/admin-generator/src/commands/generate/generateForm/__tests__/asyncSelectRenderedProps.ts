import { buildSchema, introspectionFromSchema } from "graphql";

import { type FormConfig, type FormFieldConfig } from "../../generate-command";
import { generateFormField } from "../generateFormField";

describe("AsyncSelect rendered props", () => {
    it("generates readonly async select", async () => {
        const schema = buildSchema(`
            type Query {
                products: [Product]!
                categories: [Category]!
            } 

            type Product {
                id: ID!
                title: String!
                category: Category!
            }
            type Category {
                id: ID!
                name: String!
            }

            type Mutation {
                createProduct(input: ProductInput!): Product!
                updateProduct(id: ID!, input: ProductInput!): Product!
            }

            input ProductInput {
                title: String
                category: ID
            }
        `);
        type GQLCategory = {
            __typename?: "Category";
            id: string;
            name: string;
        };
        type GQLProduct = {
            __typename?: "Product";
            id: string;
            title: string;
            category: GQLCategory;
        };

        const introspection = introspectionFromSchema(schema);

        const fieldConfig: FormFieldConfig<GQLProduct> = {
            type: "asyncSelect",
            rootQuery: "categories",
            name: "category",
            readOnly: true,
        };
        const formConfig: FormConfig<GQLProduct> = {
            type: "form",
            gqlType: "Product",
            fields: [fieldConfig],
        };

        const output = generateFormField({
            gqlIntrospection: introspection,
            baseOutputFilename: "ProductForm",
            formFragmentName: "ProductFormFragment",
            config: fieldConfig,
            formConfig,
            gqlType: "Product",
        });

        expect(output.code).toMatchSnapshot();
    });
});
