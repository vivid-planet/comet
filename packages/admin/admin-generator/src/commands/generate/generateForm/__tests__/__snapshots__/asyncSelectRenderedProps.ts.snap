// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncSelect rendered props generates readonly async select 1`] = `
"<AsyncSelectField
                
                variant="horizontal"
                fullWidth
                readOnly disabled
                
                name="category"
                label={<FormattedMessage id="product.category" defaultMessage="Category" />}
                
                loadOptions={async () => {
                    const { data } = await client.query<GQLCategoriesSelectQuery, GQLCategoriesSelectQueryVariables>({
                        query: gql\`query CategoriesSelect {
                            categories {
                                nodes {
                                    id
                                    name
                                }
                            }
                        }\`
                    });
                    return data.categories.nodes;
                }}
                getOptionLabel={(option) => option.name}
                
            />"
`;
